dags/exchange_rate.py

```diff
@@ -0,0 +1,312 @@
+    context: dagster.AssetExecutionContext, config: ExchangeRateConfig
+) -> dagster.Output[dict[str, Any]]:
+    """
+    Fetches exchange rates from the Open Exchange Rates API for a specific hour.
+    The date is determined by the partition key, which is in the format %Y-%m-%d-%H:%M.
+    """
+    # Convert hourly partition key to daily format because we always fetch information for the day
+    date_str = get_date_partition_from_hourly_partition(context.partition_key)
+    app_id = config.app_id
+    api_base_url = config.api_base_url
+
+    if not app_id:
+        raise ValueError("Open Exchange Rates API key (app_id) is required")
+
+    rates = fetch_exchange_rates(
+        context=dagster.build_op_context(), date_str=date_str, app_id=app_id, api_base_url=api_base_url
+    )
+
+    return dagster.Output(
+        value=rates,
+        metadata={
+            "date_str": date_str,
+            "rates_count": len(rates),
+            "rates": rates,
+        },
+    )
+
+
+@dagster.op
+def store_exchange_rates_in_clickhouse(
+    context: dagster.OpExecutionContext,
+    date_str: str,
+    exchange_rates: dict[str, Any],
+    cluster: dagster.ResourceParam[ClickhouseCluster],
+) -> tuple[list[dict[str, Any]], list[str]]:
+    """
+    Stores exchange rates data in ClickHouse.
+    """
+    # Transform data into rows for ClickHouse
+    rows = [
+        {"date": date_str, "currency": currency, "rate": rate}
+        for currency, rate in exchange_rates.items()
+        if currency in SUPPORTED_CURRENCY_CODES
+    ]
+
+    # Log information about the data being stored
+    context.log.info(f"Storing {len(rows)} exchange rates for {date_str} in ClickHouse")
+
+    # Prepare values for batch insert
+    # Use toDate() to cast the string date to a ClickHouse Date type
+    values = [f"('{row['currency']}', {row['rate']}, toDate('{row['date']}'))" for row in rows]
+
+    # Execute the insert if there are values to insert
+    if values:
+
+        def insert(client: Client):
+            try:
+                insert_sql = (
+                    f"INSERT INTO {EXCHANGE_RATE_TABLE_NAME} (currency, rate, start_date) VALUES {', '.join(values)}"
+                )
+
+                # Batch insert all values
+                client.execute(insert_sql)
+                context.log.info("Successfully inserted exchange rates")
+            except Exception as e:
+                context.log.warning(f"Failed to insert exchange rates: {e}")
+
+        def reload_dict(client: Client):
+            try:
+                client.execute(f"SYSTEM RELOAD DICTIONARY {EXCHANGE_RATE_DICTIONARY_NAME}")
+                context.log.info("Successfully reloaded exchange_rate_dict dictionary")
+            except Exception as e:
+                context.log.warning(f"Failed to reload exchange_rate_dict dictionary: {e}")
+
+        cluster.map_hosts_by_role(insert, NodeRole.DATA).result()
+        cluster.map_hosts_by_role(reload_dict, NodeRole.DATA).result()
```

Comment:
@PostHog/clickhouse I need some help here. I have several questions:
1. Is this the best way to do this?
2. Am I right in running this on Data nodes only or do I need to run in the coordinator as well?
3. We are not handling errors that well here, how should I handle them?
4. Does this even make sense?

Reply:
Ad 2. dictionary must be on coordinator too, cause it may be needed by it. ClickHouse coordinator may do a finishing processing / aggregating etc. of a query

Reply:
Cool, will change that @orian. Do I need to insert the exchange rates on all nodes as well or is it enough to insert it in a single table and rely on the replication?

Reply:
Changed to insert and refresh dict on all nodes, also added some error handling here